src/server/tiering/disk_storage.cc

```diff
@@ -0,0 +1,66 @@
+// Copyright 2024, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#include "server/tiering/disk_storage.h"
+
+#include "base/io_buf.h"
+#include "base/logging.h"
+#include "server/error.h"
+
+namespace dfly::tiering {
+
+std::error_code DiskStorage::Open(std::string_view path) {
+  RETURN_ON_ERR(io_mgr_.Open(path));
+  alloc_.AddStorage(0, io_mgr_.Span());
+  return {};
+}
+
+void DiskStorage::Close() {
+  io_mgr_.Shutdown();
+}
+
+void DiskStorage::Read(DiskSegment segment, ReadCb cb) {
+  DCHECK_GT(segment.length, 0u);
+
+  uint8_t* buf = new uint8_t[segment.length];
+  auto io_cb = [cb, buf, segment](int res) {
+    cb(std::string_view{reinterpret_cast<char*>(buf), segment.length});
+    delete buf;  // because std::function needs to be copyable, unique_ptr can't be used
+  };
+  io_mgr_.ReadAsync(segment.offset, {buf, segment.length}, std::move(io_cb));
+}
+
+void DiskStorage::MarkAsFree(DiskSegment segment) {
+  DCHECK_GT(segment.length, 0u);
+  alloc_.Free(segment.offset, segment.length);
+}
+
+std::error_code DiskStorage::Stash(io::Bytes bytes, StashCb cb) {
+  DCHECK_GT(bytes.length(), 0u);
+
```

Comment:
please dcheck on length 4k alignment

Reply:
you mean offset

Reply:
i meant `bytes.length()`

Reply:
> i meant bytes.length()

Didn't we discuss above what the potential benefits of 4k alignment would be if our abstractions both ways work with byte ranges? If it's always 4k bytes, then why not count pages instead 😆